File: frontend/src/services/reservaService.ts
import { api } from '@/lib/api';
import type {
  Reserva,
  AreaDisponible,
  ReservaRequest,
  ReservaAprobacionRequest,
  ReservaEstadisticas,
  FiltrosReserva,
  PaginacionReserva
} from '../types/reserva';

export class ReservaService {
  private baseUrl = '/api/reservas';

  // Obtener todas las reservas con filtros y paginación
  async getReservas(filtros?: FiltrosReserva): Promise<PaginacionReserva> {
    const params = new URLSearchParams();
    
    if (filtros?.estado) params.append('estado', filtros.estado);
    if (filtros?.tipo_reserva) params.append('tipo_reserva', filtros.tipo_reserva);
    if (filtros?.fecha_reserva) params.append('fecha_reserva', filtros.fecha_reserva);
    if (filtros?.fecha_reserva_desde) params.append('fecha_reserva__gte', filtros.fecha_reserva_desde);
    if (filtros?.fecha_reserva_hasta) params.append('fecha_reserva__lte', filtros.fecha_reserva_hasta);
    if (filtros?.area_comun) params.append('area_comun', filtros.area_comun.toString());
    if (filtros?.residente) params.append('residente', filtros.residente.toString());
    if (filtros?.search) params.append('search', filtros.search);

    const response = await api.get(`${this.baseUrl}/?${params.toString()}`);
    return response.data;
  }

  // Obtener una reserva específica
  async getReserva(id: number): Promise<Reserva> {
    const response = await api.get(`${this.baseUrl}/${id}/`);
    return response.data;
  }

  // Crear una nueva reserva
  async crearReserva(reserva: ReservaRequest): Promise<Reserva> {
    const response = await api.post(`${this.baseUrl}/`, reserva);
    return response.data;
  }

  // Actualizar una reserva
  async actualizarReserva(id: number, reserva: Partial<ReservaRequest>): Promise<Reserva> {
    const response = await api.put(`${this.baseUrl}/${id}/`, reserva);
    return response.data;
  }

  // Eliminar una reserva
  async eliminarReserva(id: number): Promise<void> {
    await api.delete(`${this.baseUrl}/${id}/`);
  }

  // Obtener áreas disponibles para una fecha
  async getAreasDisponibles(fecha: string): Promise<AreaDisponible[]> {
    const response = await api.get(`${this.baseUrl}/disponibles/?fecha=${fecha}`);
    return response.data;
  }

  // Aprobar una reserva
  async aprobarReserva(id: number): Promise<Reserva> {
    const response = await api.post(`${this.baseUrl}/${id}/aprobar/`);
    return response.data;
  }

  // Rechazar una reserva
  async rechazarReserva(id: number, observaciones?: string): Promise<Reserva> {
    const data: ReservaAprobacionRequest = {
      accion: 'rechazar',
      ...(observaciones && { observaciones })
    };
    const response = await api.post(`${this.baseUrl}/${id}/rechazar/`, data);
    return response.data;
  }

  // Cancelar una reserva
  async cancelarReserva(id: number, motivo?: string): Promise<Reserva> {
    const response = await api.post(`${this.baseUrl}/${id}/cancelar/`, { motivo });
    return response.data;
  }

  // Completar una reserva
  async completarReserva(id: number): Promise<Reserva> {
    const response = await api.post(`${this.baseUrl}/${id}/completar/`);
    return response.data;
  }

  // Obtener próximas reservas
  async getProximasReservas(): Promise<Reserva[]> {
    const response = await api.get(`${this.baseUrl}/proximas/`);
    return response.data;
  }

  // Obtener estadísticas de reservas
  async getEstadisticas(): Promise<ReservaEstadisticas> {
    const response = await api.get(`${this.baseUrl}/estadisticas/`);
    return response.data;
  }

  // Obtener mis reservas (para residentes)
  async getMisReservas(): Promise<PaginacionReserva> {
    const response = await api.get(`${this.baseUrl}/mis_reservas/`);
    return response.data;
  }

  // Validar disponibilidad de horario
  async validarDisponibilidad(
    areaComunId: number,
    fecha: string,
    horaInicio: string,
    horaFin: string,
    reservaId?: number
  ): Promise<boolean> {
    try {
      const areas = await this.getAreasDisponibles(fecha);
      const area = areas.find(a => a.id === areaComunId);
      
      if (!area) return false;
      
      // Verificar si hay conflictos de horario
      const horarioInicio = horaInicio;
      const horarioFin = horaFin;
      
      for (const horarioOcupado of area.horarios_ocupados) {
        const [inicioOcupado, finOcupado] = horarioOcupado.split(' - ');
        
        // Verificar que ambos valores existen
        if (!inicioOcupado || !finOcupado) continue;
        
        // Verificar solapamiento
        if (horarioInicio < finOcupado && horarioFin > inicioOcupado) {
          return false;
        }
      }
      
      return true;
    } catch (error) {
      console.error('Error validando disponibilidad:', error);
      return false;
    }
  }

  // Calcular costo de reserva
  calcularCosto(montoHora: number, horaInicio: string, horaFin: string): number {
    const inicio = new Date(`2000-01-01T${horaInicio}`);
    const fin = new Date(`2000-01-01T${horaFin}`);
    const duracionHoras = (fin.getTime() - inicio.getTime()) / (1000 * 60 * 60);
    return montoHora * duracionHoras;
  }

  // Obtener horarios disponibles para un área en una fecha
  async getHorariosDisponibles(areaComunId: number, fecha: string): Promise<string[]> {
    const areas = await this.getAreasDisponibles(fecha);
    const area = areas.find(a => a.id === areaComunId);
    
    if (!area) return [];
    
    // Generar horarios de 6:00 a 23:00 cada hora
    const horarios: string[] = [];
    for (let hora = 6; hora <= 23; hora++) {
      const horario = `${hora.toString().padStart(2, '0')}:00`;
      horarios.push(horario);
    }
    
    // Filtrar horarios ocupados
    const horariosDisponibles = horarios.filter(horario => {
      const hora = horario.split(':')[0];
      if (!hora) return false;
      
      const horarioFin = `${(parseInt(hora) + 1).toString().padStart(2, '0')}:00`;
      
      return !area.horarios_ocupados.some(ocupado => {
        const [inicioOcupado, finOcupado] = ocupado.split(' - ');
        
        // Verificar que ambos valores existen
        if (!inicioOcupado || !finOcupado) return false;
        
        return horario < finOcupado && horarioFin > inicioOcupado;
      });
    });
    
    return horariosDisponibles;
  }
}

// Instancia singleton del servicio
export const reservaService = new ReservaService();

File: mobile/lib/services/reserva_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/reserva.dart';
import '../config/app_config.dart';
import 'auth_service.dart';

class ReservaService {
  String get baseUrl => '${AppConfig.apiBaseUrl}/api';
  final AuthService _authService = AuthService();

  // Headers para las peticiones autenticadas
  Future<Map<String, String>> _getHeaders() async {
    final token = await _authService.getToken();
    return {
      'Content-Type': 'application/json',
      'Authorization': 'Bearer $token',
    };
  }

  // Obtener todas las reservas del usuario
  Future<ApiResponse<List<Reserva>>> getMisReservas() async {
    try {
      final headers = await _getHeaders();
      final response = await http.get(
        Uri.parse('$baseUrl/reservas/mis_reservas/'),
        headers: headers,
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        final List<dynamic> results = data['results'] ?? [];
        final reservas = results.map((json) => Reserva.fromJson(json)).toList();
        
        return ApiResponse<List<Reserva>>(
          success: true,
          data: reservas,
          message: 'Reservas obtenidas exitosamente',
        );
      } else {
        final errorData = json.decode(response.body);
        return ApiResponse<List<Reserva>>(
          success: false,
          error: errorData['error'] ?? 'Error al obtener reservas',
        );
      }
    } catch (e) {
      return ApiResponse<List<Reserva>>(
        success: false,
        error: 'Error de conexión: $e',
      );
    }
  }

  // Obtener áreas disponibles para una fecha
  Future<ApiResponse<List<AreaDisponible>>> getAreasDisponibles(DateTime fecha) async {
    try {
      final headers = await _getHeaders();
      final fechaStr = fecha.toIso8601String().split('T')[0];
      final response = await http.get(
        Uri.parse('$baseUrl/reservas/disponibles/?fecha=$fechaStr'),
        headers: headers,
      );

      if (response.statusCode == 200) {
        final List<dynamic> data = json.decode(response.body);
        final areas = data.map((json) => AreaDisponible.fromJson(json)).toList();
        
        return ApiResponse<List<AreaDisponible>>(
          success: true,
          data: areas,
          message: 'Áreas disponibles obtenidas exitosamente',
        );
      } else {
        final errorData = json.decode(response.body);
        return ApiResponse<List<AreaDisponible>>(
          success: false,
          error: errorData['error'] ?? 'Error al obtener áreas disponibles',
        );
      }
    } catch (e) {
      return ApiResponse<List<AreaDisponible>>(
        success: false,
        error: 'Error de conexión: $e',
      );
    }
  }

  // Crear una nueva reserva
  Future<ApiResponse<Reserva>> crearReserva(ReservaRequest request) async {
    try {
      final headers = await _getHeaders();
      final response = await http.post(
        Uri.parse('$baseUrl/reservas/'),
        headers: headers,
        body: json.encode(request.toJson()),
      );

      if (response.statusCode == 201) {
        final data = json.decode(response.body);
        final reserva = Reserva.fromJson(data);
        
        return ApiResponse<Reserva>(
          success: true,
          data: reserva,
          message: 'Reserva creada exitosamente',
        );
      } else {
        final errorData = json.decode(response.body);
        return ApiResponse<Reserva>(
          success: false,
          error: errorData['error'] ?? 'Error al crear reserva',
        );
      }
    } catch (e) {
      return ApiResponse<Reserva>(
        success: false,
        error: 'Error de conexión: $e',
      );
    }
  }

  // Obtener una reserva específica
  Future<ApiResponse<Reserva>> getReserva(int id) async {
    try {
      final headers = await _getHeaders();
      final response = await http.get(
        Uri.parse('$baseUrl/reservas/$id/'),
        headers: headers,
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        final reserva = Reserva.fromJson(data);
        
        return ApiResponse<Reserva>(
          success: true,
          data: reserva,
          message: 'Reserva obtenida exitosamente',
        );
      } else {
        final errorData = json.decode(response.body);
        return ApiResponse<Reserva>(
          success: false,
          error: errorData['error'] ?? 'Error al obtener reserva',
        );
      }
    } catch (e) {
      return ApiResponse<Reserva>(
        success: false,
        error: 'Error de conexión: $e',
      );
    }
  }

  // Actualizar una reserva
  Future<ApiResponse<Reserva>> actualizarReserva(int id, ReservaRequest request) async {
    try {
      final headers = await _getHeaders();
      final response = await http.put(
        Uri.parse('$baseUrl/reservas/$id/'),
        headers: headers,
        body: json.encode(request.toJson()),
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        final reserva = Reserva.fromJson(data);
        
        return ApiResponse<Reserva>(
          success: true,
          data: reserva,
          message: 'Reserva actualizada exitosamente',
        );
      } else {
        final errorData = json.decode(response.body);
        return ApiResponse<Reserva>(
          success: false,
          error: errorData['error'] ?? 'Error al actualizar reserva',
        );
      }
    } catch (e) {
      return ApiResponse<Reserva>(
        success: false,
        error: 'Error de conexión: $e',
      );
    }
  }

  // Cancelar una reserva
  Future<ApiResponse<Reserva>> cancelarReserva(int id, {String? motivo}) async {
    try {
      final headers = await _getHeaders();
      final body = motivo != null ? json.encode({'motivo': motivo}) : '{}';
      final response = await http.post(
        Uri.parse('$baseUrl/reservas/$id/cancelar/'),
        headers: headers,
        body: body,
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        final reserva = Reserva.fromJson(data);
        
        return ApiResponse<Reserva>(
          success: true,
          data: reserva,
          message: 'Reserva cancelada exitosamente',
        );
      } else {
        final errorData = json.decode(response.body);
        return ApiResponse<Reserva>(
          success: false,
          error: errorData['error'] ?? 'Error al cancelar reserva',
        );
      }
    } catch (e) {
      return ApiResponse<Reserva>(
        success: false,
        error: 'Error de conexión: $e',
      );
    }
  }

  // Eliminar una reserva
  Future<ApiResponse<void>> eliminarReserva(int id) async {
    try {
      final headers = await _getHeaders();
      final response = await http.delete(
        Uri.parse('$baseUrl/reservas/$id/'),
        headers: headers,
      );

      if (response.statusCode == 204) {
        return ApiResponse<void>(
          success: true,
          message: 'Reserva eliminada exitosamente',
        );
      } else {
        final errorData = json.decode(response.body);
        return ApiResponse<void>(
          success: false,
          error: errorData['error'] ?? 'Error al eliminar reserva',
        );
      }
    } catch (e) {
      return ApiResponse<void>(
        success: false,
        error: 'Error de conexión: $e',
      );
    }
  }

  // Obtener próximas reservas
  Future<ApiResponse<List<Reserva>>> getProximasReservas() async {
    try {
      final headers = await _getHeaders();
      final response = await http.get(
        Uri.parse('$baseUrl/reservas/proximas/'),
        headers: headers,
      );

      if (response.statusCode == 200) {
        final List<dynamic> data = json.decode(response.body);
        final reservas = data.map((json) => Reserva.fromJson(json)).toList();
        
        return ApiResponse<List<Reserva>>(
          success: true,
          data: reservas,
          message: 'Próximas reservas obtenidas exitosamente',
        );
      } else {
        final errorData = json.decode(response.body);
        return ApiResponse<List<Reserva>>(
          success: false,
          error: errorData['error'] ?? 'Error al obtener próximas reservas',
        );
      }
    } catch (e) {
      return ApiResponse<List<Reserva>>(
        success: false,
        error: 'Error de conexión: $e',
      );
    }
  }
}

// Clase para manejar respuestas de la API
class ApiResponse<T> {
  final bool success;
  final T? data;
  final String? error;
  final String? message;

  ApiResponse({
    required this.success,
    this.data,
    this.error,
    this.message,
  });
}
